docs/developer/design/authentication.rst

```diff
@@ -0,0 +1,154 @@
+connectors/plugins and it's integrated support for other
+External IDP Integration as an Identity Provider.
+
+Design choices
+^^^^^^^^^^^^^^
+
+Dex is chosen as an Identity Provider in MetalK8s based on the above
+requirements enforced by the following reasons:
+
+* Dex's support for multiple plugins greatly increases the potential for
+  integrating with existing user management system such as Active Directory,
+  LDAP, SAML and others.
+* The ability for Dex to issue tokens, revoke tokens, refresh tokens and even
+  control the lifetime of these tokens makes **Dex** suitable since we can
+  force users to re-authenticate when need arises.
+* Dex can be seamlessly deployed in a Kubernetes cluster.
+* Dex provides access to a highly customizable UI which is a step closer to
+  good user experience which we advocate for.
+* Dex provides support for multiple backends for storage. This guarantees user
+  account information persistence after upgrades.
+
+Rejected design choices
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Static passwords
+~~~~~~~~~~~~~~~~
+
+This involves adding new users by updating a static file located on every
+control-plane Node and then restarting the API server.
+
+This was rejected since there is no support for a pluggable user management
+system and it is cumbersome to operate.
+
+X.509 certificates
+~~~~~~~~~~~~~~~~~~
+
+Here, each user owns a signed certificate that is validated by the API
+server.
+
+This approach is not user-friendly meaning each certificate has to be manually
+signed. Also providing certificates for accessing the MetalK8s UI needs
+much more effort.
+
+Keycloak
+~~~~~~~~
+
+Considered as one of the heavyweight identity management systems offering a
+full blown out Identity Management.
+
+Not suitable for our use case since it entails much efforts in terms of
+deployment and management. Also since we drive for a simplistic approach to
+Identity Management, Keycloak falls short in this aspect.
+
+Keycloak is considered a future fall-back Identity Provider if the need ever
+arises from our customers.
+
+Unexploited choices
+~~~~~~~~~~~~~~~~~~~
+* ORY Hydra
+
+
+Implementation Details
+----------------------
+
+Implementation steps(High level)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Iteration 1
+~~~~~~~~~~~
+
+- Using Salt, generate self-signed certificates needed for Dex deployment
+- Deploy Dex in MetalK8s from the official **Dex Charts** while making use of
+  the generated certificates above
+- Configure Kubernetes API server flags to use Dex
+- Expose Dex on the control plane using Ingress
+
+Iteration 2
+~~~~~~~~~~~
+
+- Implement a UI integration with Dex.
+- Theme the Dex GUI to match MetalK8s UI specs.
+- Provide documentation on how to create Dex users and secrets.
+
+Iteration 3
+~~~~~~~~~~~
+
+- Provide CLI Integration to create/manage users and corresponding
+  documentation.
+- Provide documentation on how to integrate with an external Identity Provider
+  focusing on Active Directory and LDAP.
+
```

Comment:
Shouldn't we split those iterations so that each one is bringing some value and is potentially demo-able and shippable to production?